x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/AggregationResultUtils.java

```diff
@@ -334,6 +340,27 @@ public Object value(Aggregation agg, Map<String, String> fieldTypeMap, String lo
         }
     }
 
+    static class RangeAggExtractor implements AggValueExtractor {
+        @Override
+        public Object value(Aggregation agg, Map<String, String> fieldTypeMap, String lookupFieldPrefix) {
+            Range aggregation = (Range) agg;
+            HashMap<String, Long> ranges = new HashMap<>();
+            for (Range.Bucket bucket : aggregation.getBuckets()) {
+                double from = bucket.getFrom() == null ? Double.NEGATIVE_INFINITY : ((Double) bucket.getFrom()).doubleValue();
+                double to = bucket.getTo() == null ? Double.POSITIVE_INFINITY : ((Double) bucket.getTo()).doubleValue();
+                ranges.put(generateKeyForRange(from, to), bucket.getDocCount());
+            }
+            return ranges;
+        }
+    }
+
+    static String generateKeyForRange(double from, double to) {
+        return new StringBuilder().append(Double.isInfinite(from) ? "*" : OutputFieldNameConverter.fromDouble(from))
+            .append("-")
+            .append(Double.isInfinite(to) ? "*" : OutputFieldNameConverter.fromDouble(to))
+            .toString();
+    }
```

Comment:
Something that just came to mind is: How do we support multi-bucket  aggs when they are passed `keyed: false`?

It would be cool if our handling of range was easily extensible to `ip_range`, `date_range`, `geo_distance`, but I suppose this is an OK first pass.

Reply:
> My main concern is: Does having * in a field name cause issues with subsequent aggs or queries against those fields? I would assume not because we are allowed to create fields like that. But, just being a little paranoid :D

Yes, `*` in the field names seems strange but nothing failed in the integration tests and manual tests I made so far.

> How do we support multi-bucket aggs when they are passed keyed: false?

IIUC regardless of the value of `keyed` we index the documents into destination index in the same format i.e. a key-value map rather than a list.
I've added a few assertions to `TransformPivotRestIT.testPivotWithRanges` that verify it is actually the case for `range`.

> It would be cool if our handling of range was easily extensible to ip_range, date_range, geo_distance, but I suppose this is an OK first pass.

+1
ip_range, date_range, geo_distance are not supported in this PR but hopefully the logic will be reusable.